src/Analysis/Ast/Impl/Types/PythonClassType.cs

```diff
@@ -281,5 +284,55 @@ internal PythonClassType(string name, Location location)
         public bool Equals(IPythonClassType other)
             => Name == other?.Name && DeclaringModule.Equals(other?.DeclaringModule);
 
+        private IEnumerable<IPythonType> DisambiguateBases(IEnumerable<IPythonType> bases, IExpressionEvaluator eval) {
+            if (bases == null) {
+                return Array.Empty<IPythonType>();
+            }
+
+            if (eval == null) {
+                return FilterCircularBases(bases).Where(b => !b.IsUnknown());
+            }
+
+            var newBases = new List<IPythonType>();
+            foreach (var b in bases) {
+                var imported = eval.LookupImportedNameInScopes(b.Name, out _);
+                if (imported is IPythonType importedType) {
+                    // Variable with same name as the base was imported.
+                    // If there is also a local declaration, we need to figure out which one wins.
+                    var declared = eval.LookupNameInScopes(b.Name, out var scope);
+                    if (declared != null && scope != null) {
+                        var v = scope.Variables[b.Name];
+                        if (v.Source != VariableSource.Import && v.Value is IPythonClassType cls && cls.Location.IndexSpan.Start >= Location.IndexSpan.Start) {
+                            // There is a declaration with the same name, but it appears later in the module. Use the import.
+                            if (!importedType.IsUnknown()) {
+                                newBases.Add(importedType);
+                            }
+                            continue;
+                        }
+                    }
+                }
+
+                if (!b.IsUnknown()) {
+                    newBases.Add(b);
+                }
+            }
+            return FilterCircularBases(newBases);
+        }
+
+        private IEnumerable<IPythonType> FilterCircularBases(IEnumerable<IPythonType> bases) {
+            // Inspect each base chain and exclude bases that chains to this class.
+            foreach(var b in bases.Where(x => !Equals(x))) {
```

Comment:
So this is dropping bases that have chains that lead back to the original class type. Didn't we want to save those bases to process them later? If we don't, then do we still reentrancy guards?

Reply:
Nah, we don't want to process loops. Depends on guards. MRO guard might still apply...

Reply:
Hmm perhaps the generic resolution guard is unnecessary now. Not sure about mro, I don't see how it could get loops if all the bases with loops are removed

Reply:
Maybe. But I'll leave it now for safety. We'll see f it can be removed later.

Reply:
OK